src/dune_engine/build_config.ml

```diff
@@ -11,16 +11,36 @@ module Context_or_install = struct
     | Context s -> Context_name.to_dyn s
 end
 
-type rules =
-  { build_dir_only_sub_dirs : Subdir_set.t
-  ; directory_targets : Loc.t Path.Build.Map.t
-  ; rules : Rules.t Memo.t
-  }
+module Rules = struct
+  type t =
+    { build_dir_only_sub_dirs : Subdir_set.t
+    ; directory_targets : Loc.t Path.Build.Map.t
+    ; rules : Rules.t Memo.t
+    }
+
+  let empty =
+    { build_dir_only_sub_dirs = Subdir_set.empty
+    ; directory_targets = Path.Build.Map.empty
+    ; rules = Memo.return Rules.empty
+    }
+
+  let combine_exn r { build_dir_only_sub_dirs; directory_targets; rules } =
+    { build_dir_only_sub_dirs =
+        Subdir_set.union r.build_dir_only_sub_dirs build_dir_only_sub_dirs
```

Comment:
By the way, is this `Subdir_set.union` really `Subdir_set.union_exn`? If yes, maybe we should rename it so we have consistent naming.

Reply:
Not in this case, because `Subdir_set.union` cannot raise. Here's the implementation:

```
let union a b =
  match (a, b) with
  | All, _ | _, All -> All
  | These a, These b -> These (String.Set.union a b)
```